### ISSUE TITLE
Usability Discussion

### ISSUE THREAD

**Issue by TheLarkInn** [webpack/webpack#2797]
In light of much conversation on twitter, etc. and seeing the many "flavors" of abstraction layers or utilities on top of webpack, I believe that it is probably in our best interest to bring these authors together so that we can come up with some incredible ideas and potentially draft a proposal that will allow webpack to be easier to use out of the box for any web developer. 

**Problems to solve:** 
"Webpack is too low level"
"There are no reasonable set of defaults"
"Terminology doesn't make sense"
"Too much boilerplate in handling environment logic"
"The CLI lacks scaffolding and initialization"
"Configuration is 'convoluted'"

**Things to keep into consideration:** 
* Webpack is unopinionated and makes no assumptions about a development workflow
* Webpack is framework/library agnostic and so any defaults must be able to be pragmatic enough to cater to any library. 

So this is a discussion group open to anyone but tagged with some people in mind whom I feel have already great tools created, or great ideas. However, is open for discussion by anyone.

**Comment by TheLarkInn** [comment]
CC: 
@kentcdodds @ericclemmons @joshhunt @gaearon @mxstbr @bebraw @jhnns @gdi2290

**Comment by TheLarkInn** [comment]
CC 
@niieani

**Comment by TheLarkInn** [comment]
I suppose I'll be the first to start-off why not. 😏 

I think overall IDE support would be awesome for validation. 

I'm not for forcing people to use Typescript, however there are some wonderful automatic validations come when you write Webpack Configs with Webpack Typings and Typescript. It reminds me of `webpack-validator` but automagical in your IDE. Typings are already being tackled as we speak for webpack 2.x. Also a pragmatic set of defaults and a `config` generator are all good places to start at.

**Comment by d3viant0ne** [comment]
@TheLarkInn - Could we not easily find a leverage `webpack-validator` in the same manner that Angular2 projects are  leveraging `codelyzer` and let an existing and mature linting framework ( eslint / tslint out of the gate) do the heavy lifting? 

That would handle ES6 & TypeScript in all the major IDEs out of the gate and given people that "automagical" feedback as well as explicit execution for things like CI validation.

**Comment by joshhunt** [comment]
For me, 'linting/validation' and 'usability concerns' are two seperate, but related, things. I don't think flagging errors in a Webpack config does quite enough to make it easier to use out of the box.

I think there are a bunch of low hanging fruit in the config format itself that could be address to make the most common use-cases easier a bit simpler. I'm not sure what the best approach is, but I guess it'll depend on the tradeoff with compatibility.

From the [getting started][1] guide and [CSS Loader readme][2], a relatively simple Webpack config would look like:

```
module.exports = {
    entry: "./entry.js",
    output: {
        path: __dirname,
        filename: "bundle.js"
    },
    module: {
        loaders: [
            { test: /\.css$/, loader: "style-loader!css-loader" },
            { test: /\.png$/, loader: "url-loader?limit=100000" },
        ]
    }
};
```

To me, most of the confusing parts of Webpack are in the loader configuration, with both the regex `test` and the weird 'query string' format for passing in loaders and their configuration. Considering the most common use case would be adding a loader for a certain file type, I think it can be optimised for this:

```
module.exports = {
    entry: './entry.js',
    output: './bundle.js',
    loaders: {
        'css': ['style', 'css'],
        'png': [{ loader: 'url', limit: 100 }]
    }
};
```

This is more or less compatible with configs and loaders that are out there at the moment. What I don't like about this is although it simplifies this config, supporting both formats would make everything more confusing overall.

I think the best approach (for ease-of-use) would be to start from a complete blank slate, as explored in ericclemmons/terse-webpack and joshhunt/nicepack. Is this on the cards?

[1]: https://webpack.github.io/docs/tutorials/getting-started/#config-file
[2]: https://github.com/webpack/css-loader#example-config

**Comment by joshhunt** [comment]
For what it's worth, I think Webpack should stay agnostic to any underlying technology (no babel presets), but understand what the common use cases are, like adding something to transform all scripts from CoffeeScript/Babel/TypeScript.

**Comment by bebraw** [comment]
Based on my experience with the tool, the fact that you have multiple ways to configure something like loaders can lead to confusion. You need to understand `loader` vs. `loaders` and all those query formats. Even if you stick with `loaders`, you still have to use `loader` with `ExtractTextPlugin.extract` due to the way it works.

If we can provide a single, clear way to achieve things like loader configuration on core level, that would go a long way. You can build tools like `terse-webpack` on top of it. From my perspective it would be ideal to have something that's clear to explain.

## On Schema

`webpack-validator` is actually quite nice and helps quite a bit. I do feel though that configuration schema is a core concern as that enables tool authors to do all sort of things. And if schema was maintained at core level, then `webpack-validator` wouldn't have to maintain it. And I expect you could build better IDE integration, and even web services and better documentation, on top of it.

A schema would help with one of my favorite pet peeves of webpack, the CLI. A lot of support requests gravitate around poor understanding of the CLI and what it does. Often you see people enabling HMR twice accidentally. There are some inconsistencies around that don't particularly help.

I feel there could be room for simplification and I might go personally as far as strip the CLI from webpack core and let community maintain it at least for a while. Maybe a good version could be blessed and moved to be maintained within the official organization.

The core should make it easy to write a CLI (schema!) and provide a solid Node API. If it can achieve those at a high level, it would improve the situation substantially.

**Comment by niieani** [comment]
Author of [Easy Webpack](https://github.com/easy-webpack/core) here.

For me, the biggest problem with maintaining raw Webpack configuration files was the separation of concerns.
You couldn't easily split your config file into sections like "TypeScript", "production-specific", "development-specific", "Electron-specific", etc., because there is no built-in, pain-free way to merge them afterwards. The reason is that some configuration elements are nested Objects and some nested Arrays. In 99% of cases the solution is to: "deep-merge" the Objects and "shallow-append" Arrays, but sometimes you'd like to ensure a certain plugin is at the end (e.g. the `offline-plugin` recommends it) and sometimes you'd like to base your config values on the previous one.

The main goal of Easy Webpack was to make process this pain-free. (note: there's also @bebraw's [webpack-merge](https://www.npmjs.com/package/webpack-merge) that has a similar aim). 

After separation of concerns is addressed, it becomes much easier for users to reason about their configs. They know "what part does what and belongs to what" and they can keep their configs DRY when they need alternative ones for all the different environments: production, development, electron backend, electron frontend, testing, ...

Suddenly, config files become about "composition of features" rather than a mess of low-level concepts (loaders, regexes, plugins, devtools, targets, ...)  that the users otherwise needed to know about.

This has lead me to the second idea - that these "features" could then be made into NPM-presets that users can just `npm install`, and stack into their configs.
The reasoning is simple: We all know **best-practices change all the time**. Loaders and Plugins evolve, get abandoned, get forked, get breaking changes in the way they're configured. It's almost a full time job getting up to speed with all of this, even if in 90% of the cases you just want the basic functionality. Ultimately, nobody wants to spend time in their configuration files - it's "work-about-work", not actual, productive **work**.

I thought that it would be great to be able to "offload" the responsibility of knowing "the how" of implementation of a certain feature (like TypeScript loading and checking) to the developers that really know what they're doing, and thus created [a number of NPM packages](https://www.npmjs.com/~easy-webpack) that contain both the dependencies and sane defaults for configs, which are additionally configurable for the most common use-cases. The config templates can be manually overwritten, so the user still has full control. 

The result? I'm the Core Team Member of [Aurelia](http://aurelia.io), one of my responsibilities is Webpack. The users who never tried Webpack were all dazed and confused by the sheer length, complexity and number of Webpack config files present in our previous Starter Kit - and Easy Webpack has worked wonderfully. I was even able to push fixes to the configuration files that didn't require us to write long blog posts explaining "you need to change this part here, and here, and replace this". All they had to do was `npm update`.

Since @TheLarkInn mentioned it, I'll address TypeScript. The beauty of using it for config files is not only built-in validation, but even more importantly, **auto-completion / auto-suggestions** and documentation -- right there where you need them the most. No need to jump to README's and websites to find out "what's the name of that property", or "what can I pass here", "can this also be a function? - wow!". Since typings are being handled, this will solve a part of the issue. 

Another win of having typings is that some IDEs are able to **enhance pure JS scenarios** and offer partial auto-suggestions anyway if they are present. These include VSCode, VS, WebStorm and I believe Atom has a plugin. To start using config typings for Webpack configs, I've [prepared these](https://github.com/easy-webpack/core/blob/master/src/webpack.ts) as a part of my project.

Yesterday, I became aware of @ericclemmons' [terse-webpack](https://github.com/ericclemmons/terse-webpack). 
I really like the fluent API - if it were written in TypeScript or had typings I'd consider it a much better API than Easy Webpack's for the end-user, since you wouldn't even need to know what presets are available - you'd see it as you're typing :). Ultimately, it could be yet another abstraction layer though, but I believe it should be possible to combine it with configuration stacking, so that we're not leaving out those who really need custom solutions.

To summarize, I think it would be wonderful to cover these layers of the onion:
1. better IDE support for configuration (docs-as-I'm-writing, autocompletion, validation)
2. separation of concerns - config-merging
3. NPM preset's -- "features, not configs" - abstracted, opinionated, higher level config packages with required dependencies for the most common use-cases -- for the 90% of users who simply want to "add that feature and be done with it". The additional benefit is that the answer to the question: "what did I need that package for in `package.json`?" [becomes clearer](https://github.com/aurelia/skeleton-navigation/blob/master/skeleton-typescript-webpack/package.json#L91-L108).
4. fluent APIs for the presets -- auto-installation and auto-configuration of the above presets in a fluent manner like terse-webpack.

🔥

**Comment by mxstbr** [comment]
I agree with all the things @bebraw said, especially

> If we can provide a single, clear way to achieve things like loader configuration on core level, that would go a long way.

This is really important!

**Comment by ericclemmons** [comment]
_(Bolded for easier skimming)_

### tldr;
- My peers greatly **prefer "Gulp vs. Grunt" configuration**
  due to IDE-support, helpful runtime errors, and composition.
- Webpack needs to support a **single, serializable object** somehow.
- Wrappers can still exist in user-land, but ideally those would
  be like [Bourbon][5], **simplifying complex, composed features**.
- Like when Babel 6 was released, things were harder for new users
  because of "Dependency Fatigue" ;) Auto-installing greatly reduces this.
- I'd much rather solve this for _all_ Webpack users than maintain
  another lib. Until then, I have a team of Gulp **users that could
  really use an easier Webpack, including myself.** ;)

### Behind the scenes at [@terse/webpack][0]

- Webpack **Object configs are architecturally flawed for composition**.
  - We've **made the same mistake as Grunt**, supporting multiple variations of the same thing (e.g. `entry`, `loaders`.
  - **Merging objects is hard**.
  - **Validating these variations is hard**.
  - **Inspecting plugin instances is hard**.
  - Especially for `.css`,**loader order matters**. 
    _I still haven't internalized what's the right order._
  - Features like HMR means changing both `entry` & `plugins`,
    which differ depending on `target`.

- **[terse][0]** attempts to solve this via:
  1. Functions instead of Objects.
    - **Throw run-time errors** with invalid arguments.
    - **Typed** with Flow or TypeScript for IDEs.
    - Support user-land functions to **compose more complex configs**
      (e.g. offline, HMR, long-term caching).
    - `.when(predicate, callback)` for **env-specific configuration**.
      (If you're going to go fluent, `if` statements break the flow!)

  2. Fluent API is just a wrapper:
    - **Builds an internal, immutable history**, much like Redux actions.
      _(I originally used Redux because people understand it now,
      but found it's kinda overkill.)_
      This is so that **merging configs is as simple as splicing arrays.**
    - Calling `getState()` will replay this history through ["feature reducers"][1] that will generate a `state` object. Doing this means there's **a single, normalized, serializable object to describe the config, which Webpack doesn't have right now.**  
    - Finally, `getConfig()` converts `state` into a Webpack config via ["webpack reducers"][2].

  3. **Installing missing deps has had the single biggest impact on my DX.**
    - Powered by [npm-install-webpack-plugin][3].
    - Even plugins are described with the intent to [auto-install][4].


[0]: https://github.com/ericclemmons/terse-webpack
[1]: https://github.com/ericclemmons/terse-webpack/tree/master/src/features
[2]: https://github.com/ericclemmons/terse-webpack/tree/master/src/reducers
[3]: https://github.com/ericclemmons/npm-install-webpack-plugin
[4]: https://github.com/ericclemmons/terse-webpack/issues/15
[5]: http://bourbon.io/

**Comment by jhnns** [comment]
This is a great discussion so far. Configuration is imho one of the biggest pain points when starting with webpack. The config style almost encourages misconfiguration and people copy wrong configuration from snippets all over stack overflow.

These are the biggest pain points I've collected so far on stack overflow:

- **Key ambiguity**: There are multiple ways to name a key (`loader` vs `loaders`)
- **Value ambiguity**: Often you can specify multiple types (strings, arrays, regex, etc.)
- **Too much possibilities to config similar things**: `modulesDirectories` vs. `root`
- **Complex string formats**: E.g. `devtool` or loader queries
- **Autoresolution**: E.g. the possibility to omit the `-loader` postfix. For instance, people get weird errors when they happen to have a [`css` folder in their `node_modules`](https://www.npmjs.com/package/css).
- **Nested objects/arrays**: The loader part is unnecessarily nested and makes merging hard
- **Suprising behavior**: E.g.: Many people don't know that every loader which `test` succeeds will be applied.


Basically we have these three options:

1. Introduce breaking changes and simpilfy config format
2. Introduce typings and linting to hint errors
3. Provide intelligent presets

I think, we should do all of them in this order. First simplify the config format, then work on tools, then provide intelligent presets. By simplifying the config format first, we remove many possible errors in the first replace, thus making our linting tools and presets a lot of easier.

In order to simplify our config format, I'd take these steps:

- **Key ambiguity**: Remove redundant config keys.
- **Value ambiguity**: We should choose the most appropiate config value and only allow this one. Primitives are good for simple configs, arrays are good when the order is important and multiple instances are allowed, objects are good when the order is not important and there should be only one option per key, functions are ultra-flexible and can be runtime dependent.
- **Too much possibilities to config similar things**: It's good that webpack is so flexible, but there are some built-in configs that many people don't need (I'd bet most people don't want the actual behavior of `modulesDirectories`). We should spot the most popular configs and make these built-in. All other use-cases should be configurable via plugins.
- **Complex string formats**: Should be simplified if possible (do people really want to configure the pragma style of the `devtool`?). We should check if they can be replaced with objects which are often easier to handle.
- **Autoresolution**: Should imho be removed. Explicit configuration is almost always better.
- **Nested objects/arrays**: Nesting should be avoided if possible.
- **Suprising behavior**: Should be removed as far as possible. In this particular case, I think the loader resolution should stop after the first `test` succeeds.

@ericclemmons I really like the idea of config reducers that translate the fluent API into a single config state object. This way we can easily catch wrong configuration on runtime.

Could you explain why the object should be serializable?

**Comment by Kovensky** [comment]
At least some of those points are solved by `webpack@2`, such as `modulesDirectories` vs `root`.

**Comment by ericclemmons** [comment]
I think @jhnns summarized everything _perfectly_.

Regarding the "serializable" part, it was to...

a. Make it easier to tell how plugins should be instantiated.
b. Provide an escape-hatch back to the raw, webpack config:

  > https://github.com/ericclemmons/terse-webpack/blob/master/src/stringify.js

If your suggestions were largely implemented, I think all of our wrappers would be largely unnecessary.

The last remaining items are around how we'd want to support/handle IDE support & types...

**Comment by jhnns** [comment]
Ok, so with serializable you don't mean it should be transformable into a string which can be written to disk? If this was the case, we could not use functions which are very handy in some cases.

I'd really love to see the terse-webpack API in webpack itself, but this is my personal opinion. I'll discuss this at our next webpack meeting. If this was ok for everyone, @ericclemmons would you be open to take the lead with our support?

Since this would be a breaking change, we really need to be careful about the changes. IMHO we may discuss about removing odd features, but if there is a valid use-case for it, it should be configurable. I think we also need to support the old format for quite a while for backwards compatibility.

**Comment by ericclemmons** [comment]
@jhnns https://github.com/yahoo/serialize-javascript does a good job of handling functions.  IME, `func.toString()` and `regex.toString()` do a great job of serializing.  I didn't consider serialization for caching to disk, but as the simplest way to make a JSON string from an object :)

Backwards compatibility is quite possible by reducing the new format to the old.  Heck, we could even do this:

```js
// Old webpack.config.js
module.exports = {
  // You know what this looks like already, otherwise we wouldn't have this open issue :D
};
```

```js
// New webpack.config.js
module.exports = require("webpack").configure({
  // strict format goes here, and is reduced internally to the current format
};
```

The fluent API _is hard to solve_ in a way that supports composition of user-land features.   I think that should exist in the future, though, as the IDE/type support would be incredibly useful IMO.

(I'm up for a Skype/Hangout, if desired.)

**Comment by TheLarkInn** [comment]
> Since this would be a breaking change, we really need to be careful about the changes. IMHO we may discuss about removing odd features, but if there is a valid use-case for it, it should be configurable. I think we also need to support the old format for quite a while for backwards compatibility.

Very true. I see this as a Webpack 3.x level breaking change. 

> The fluent API is hard to solve in a way that supports composition of user-land features. I think that should exist in the future, though, as the IDE/type support would be incredibly useful IMO.

IDE Support we could maybe start the ball rolling on this ASAP. And as we draft a new spec for this configuration. @DanielRosenwasser, @d3viant0ne, I think typings can really help this also.

**Comment by niieani** [comment]
Generally, I agree with pretty much everything that's been said here. The real issue is how to proceed with this knowledge. I'd also be up for the Skype/Hangout if that's gonna happen.

**Comment by jhnns** [comment]
@niieani @ericclemmons just an update on this:

We've talked about this [at our last webpack meeting](https://github.com/webpack/meeting-notes/pull/7/files). We could not convince everyone to use a fluent API because benefits were not clear enough. Could you summarize what the actual benefits of a fluent API are in terms of usability?

However, we all agreed that there is room to remove complexity. I'll come up with a list of unnecessary complex features/concepts which could be improved (e.g. preLoaders vs loaders vs postLoaders). Do you have any suggestions? What exactly could be removed/replaced/simplified? I'm not just talking about the actual configuration but also about concepts.

When we have a longer meeting about this topic again, we will certainly invite you.

**Comment by niieani** [comment]
@jhnns I think I'm okay with Webpack core not having a fluent API, as this could be delegated to external projects like Terse. 

What I would suggest though is Webpack at least handle config merging/composition. That shouldn't be too hard, and If that's solved without external tools, then I think it'll lead way to very nice little feature-snippets that people can than re-use across many projects, without the need to dig too deep into the docs. A fluent API can be built on top of config merging/composition from the bits. 

This is how I solve [composition](https://github.com/easy-webpack/assign/blob/master/src/index.ts) in Easy Webpack. Then there's also [webpack-merge](https://www.npmjs.com/package/webpack-merge).

Perhaps it could be as simple as accepting an array of configs to be merged? It would be nice to also be able to access the merging method programmatically though.

**Comment by gaearon** [comment]
I posted some thoughts here: https://github.com/webpack/meeting-notes/pull/7/files#r72888163. Nothing new that wasn't mentioned here before. I just think that both tooling and fluent APIs are red herrings. In my opinion the configuration should stay an object, but the *shape* of the config needs to be organized around use cases instead of "what low level options does Webpack need to know". So far Webpack configuration has been tool centric rather than user centric and this is what I think needs to be fixed.

**Comment by niieani** [comment]
I'm in total agreement with @gaearon's notes here.
Perhaps what needs to be done is finding an answer to the question: 

"How would you do configuration if you were designing a tool like Webpack in 2016?"

And then once we have a great answer, think about best ways to backport those ideas to what we have now, and eventually migrate to the design in Webpack 3.

**Comment by TheLarkInn** [comment]
@gaearon by the way you describe the use-case centric setup screams babel-like presets to me. 

No?

**Comment by mxstbr** [comment]
What's the issue with that? 

```JS
presets: ['react', 'hot', 'dev-server'] 
```

**Comment by niieani** [comment]
@mxstbr this is kind of what I've done with Easy Webpack, except that you actually have to "require" those presets in as parameters instead of typing in strings, plus those presets can have options passed into them (although babel ones can do too). It would be pretty easy to remove the `require`s and move to a string-based approach, though.

**Comment by TheLarkInn** [comment]
Where you never had to write a config, but instead in 'userland' you had a `.webpackrc` that just contained things like: 
```
{
  base: [react, es6, sourcemaps],
  
  prod: [minify, long-term-cache]
}
```

**Comment by niieani** [comment]
The reason I went with `require`s instead of a JSON file like `.webpackrc` is that you get suggestions/autocompletion and syntax validation out of the box thanks to TypeScript's typings (which also work in JavaScript in many major IDEs). 
JSON could have a schema, but that requires maintaining extra plugins to validate, they need manual management, and so on.

**Comment by gaearon** [comment]
I don't think presets would solve this. Loaders are the least confusing part of Webpack IMO even despite many ways to do the same thing.

I think file and remote path configuration (publicPath, contentBase, output.path, "name" in loaders, ?path in WDS client entry point) is the most confusing part. It is spread so thin in the configuration that it causes rippling changes all the time. It is not clear what either of these options do, and how you need to configure them to achieve common setups. And the lack of sane defaults (like filenames hashing in production, or even default `output` config that just works) forces people to learn about all those options.

**Comment by niieani** [comment]
Regarding loaders, I personally don't like that they're passed in as strings instead of constructors/instances, like plugins. It's not consistent, and it just makes it harder for novice users to grasp how things work. On top of that there's the whole "loader-resolvers" and the magical appending of "-loader" which is optional. Why all the added complexity? 
I think these sort of things should be re-thought, consolidated and simplified.

**Comment by ericclemmons** [comment]
I think the most attainable goal is to solve the "Object shape") problem as @jhnns listed them out earlier:
> https://github.com/webpack/webpack/issues/2797#issuecomment-235093740

Anything on top is just sugar, but right now **webpack configs need to be a singular, predictable format.**

**Comment by TheLarkInn** [comment]
> Regarding loaders, I personally don't like that they're passed in as strings instead of constructors/instances, like plugins.

I haven't seen which plugins internally apply loaders, but this almost _feels_ like a viable option. So do you think webpack would be easier to use if people only needed to know how to apply and use plugins? I think the abstraction is written to help prevent boilerplate the async waterfall handling of loaders. Etc. Just a thought.

**Comment by TheLarkInn** [comment]
> I just think that both tooling and fluent APIs are red herrings. In my opinion the configuration should stay an object, but the shape of the config needs to be organized around use cases instead of "what low level options does Webpack need to know". So far Webpack configuration has been tool centric rather than user centric and this is what I think needs to be fixed. 

I agree with this, but I don't believe it can be appropriately tackled until we clean up some of the issues regarding what @jhnns touched on.

**Comment by niieani** [comment]
@TheLarkInn What I meant with my comment was not to challenge the notion of loaders itself, just the fact that they need to be configured with require-like strings that are then resolved to the actual packages. On the other hand, plugins are added as actual instances. That's the inconsistency.

**Comment by mohsen1** [comment]
This might be way out of scope of this discussion but to me, any build tool that I can't simply put a breakpoint and see what's going on is flawed. Grunt, Gulp and Webpack so far been the same. Although old tools I used to use been the same (Ant 😂) I love to see a build tool that have a low level API and lets me "program" my build process instead of giving it a vague config file and hope it works fine.

**Comment by niieani** [comment]
@mohsen1 you can do that with Webpack, but it's not well documented right now; you can also use it programmatically rather than via the CLI. But it's a good point that making this debugging process easier might also be step in the right direction.

**Comment by jhnns** [comment]
@gaearon I agree. Could you post an example how a use-case centered config would look like?

> I think file and remote path configuration (publicPath, contentBase, output.path, "name" in loaders, ?path in WDS client entry point) is the most confusing part

That's interesting. I had no bigger issues with paths – but I totally agree that setting up HMR can be really frustrating.

> "How would you do configuration if you were designing a tool like Webpack in 2016?"

That sounds great. I'd really love to see code examples by all of you guys how you would like to use webpack.

---

Although I really want to simplify things, I still think that it should be possible to do that "fine-tuning" like you can do it today with webpack. Since webpack is a tool where all the different parts of the application are assembled, it's often necessary to do weird stuff just to make things work. I'm especially thinking of all the things you have to do to import modules without a common module format (which might be abstracted into a community-driven module but that's another story).

---

> Regarding loaders, I personally don't like that they're passed in as strings instead of constructors/instances, like plugins

That's true. I think it originates of CommonJS' idea of loaders (which were never implemented in node.js) where you'd configure your import by modifying the string passed to `require()`. Obviously, this is very cumbersome for stuff you import very often (like React components, CSS, HTML or images). But sometimes it is necessary to override the common behavior in this particular case (because you want to inline it with the url-loader for instance).

We had a pretty long discussion about the pros and cons of different ways to configure loaders at #378. After some years experience we know that using strings for configuration can be impossible as some tools require functions for configuration which are not serializable safely. So, personally I would be open to completely ditch the string configuration part (including all the stuff with prepending it with `!` and `-`) in favor of an external configuration via a `webpack.config.js`.

What we previously had completely uncovered is that it can be important to configure what loaders are applied on imports depending on the **importing** module. For instance, in JS you could expect to get an image just as url string, but within HTML maybe it would be better to receive a string like `src="path/to/img.jpg" width="200" height="100"` or a string referencing different versions of the same image via `srcset`. I don't know if loaders are the right tool to achieve this kind of things (essentially imports + metadata), but there's still room for improvement.

**Comment by wearhere** [comment]
Hi folks, one thing that I haven't seen come up in this thread is the difficulty of migrating from CLI usage to `webpack.config.js` usage. The CLI has wonderful "presets" like `-d` and `-p`,  and https://github.com/webpack/docs/wiki/cli describes what combinations of long flags those presets imply, but not what `webpack.config.js` options/plugins correspond to the long flags.

So what I am advocating is:
* add "development" and "production" "presets" to `webpack.config.js` too
* document how you'd reproduce each CLI flag in `webpack.config.js`

I actually found Webpack very simple and powerful to use just from the CLI. My pain came when I had to add the one option (an external, fwiw) that the CLI didn't support.

**Comment by joshfarrant** [comment]
Hi all! I'd just like to throw my two cents in regarding the config file.

The list of available config options is already quite long, and I'd imagine that it's only going to get longer. In itself this is great; more config = more flexibility = happier developers. The more options we have, however, the easier it will be to make mistakes when adding new properties.

My suggestion is that Webpack should aim to notify the user of any unrecognised properties on the config object, in the hopes of reducing time spent chasing down errors/bugs caused by incorrect property placement.

As an example; today I spent ~40 mins trying to find and fix a bug that ended up being caused by me having misplaced the `externals` object within the `output` object, rather than as a top-level config property. In hindsight this is _fairly_ clear in the docs, but at the time I was left scratching my head wondering why this solution worked for a dozen people in various examples, but wasn't working for me.

If something like this is straightforward enough to catch at run time, I think it would be useful to warn the user about it and hopefully save them a bit of time. Something along the lines of the following could work.

```
Warning: Unrecognised 'config.output.externals' property specified. Did you mean 'config.externals' ?
```

Would love to hear others' thoughts!

**Comment by andywer** [comment]
Hey there. A short, opinionated view on web config from me, too.

As a webpack user I think in dev features like *HMR*, *Babel*, *CSS Modules*, *Postcss plugins*. The webpack config instead is sliced diametrically to this: Loaders, plugins, entry points and many more. So in order to get HMR, for instance, which only affects one type of my assets (JS code), I need to edit multiple places: *dev server*, *entry points*, *js loader*.

This is not only uncomfortable for me as a user, it also makes composing a webpack config of multiple smaller configs unnecessary hard. This is why packages like `webpack-merge` exist, but trying to merge multiple loader configs identified by regexs (imagine `/jsx?/` vs. `/js|jsx/`) is pretty much doomed to fail.

And that's basically the same underlying issue @gaearon pointed out: Tool-centric vs. user-centric config. I would really love to see a feature-centric webpack config which is can safely divide into *babel config*, *hmr config*, *css modules config*, ... 😉

**Comment by andywer** [comment]
PS: After having a night's sleep...

I like @joshfarrant's proposed warning on unrecognized options. Would be a helpful feature, but suggesting the right property name might be very complex due to the deeply nested objects.

It's probably **the major problem** of today's webpack config that it looks like the under-the-hood-workings of webpack itself. This is the reason why it feels so unintuitive to most users and why you have to edit three different objects to add HMR, for instance (see my prev comment).

**Comment by ericclemmons** [comment]
@andywer That was my take as well and what my experiment was aiming to address: _what_ you wanted without worrying about _how_.

The makes Webpack incredibly powerful (and what I think will provide longevity to the project) is that it's very much about the _how_ to accomplish features, which invariably require tweaks depending on how your specific project needs it.

If it were feature based, Webpack would constantly be moving to accommodate each scenario based on how the ecosystem solved it this week, and would probably be large source of frustration for those needing to customize the config because they didn't build their app 100% the way CRA does (for example).

I really think the schema validation was a good move.

Having webpack work well with _features_, IMO, would require a herculean effort by the community to flush out all of these edge-cases in a way that worked harmoniously with every permutation of them.

I'm all for it, but don't know how to solve it _at scale_ today.  I just know what I'd like to see solved & how. And even that has been tricky :)

**Comment by TheLarkInn** [comment]
I think in terms of features: more mature plugins (yes write plugins that set config values and call other plugins) could potentially fix this. However as @ericclemmons stated the edge cases && interoperability with other plugins poses a large challenge.

**Comment by andywer** [comment]
@ericclemmons You mentioned a lot of good points, but I think the webpack config could still be straightened up to some degree without loss of configuration capabilities, just by sticking to a small set of concepts consistently.

To give you an example: Let's say every functionality of webpack is provided by a *plugin*, even the core functionalities. So you configure a set of plugin, plus your loaders, only. No more `{ devTool: 'source-map' }`, but rather `{ plugins: [ ..., webpack.devTool('source-map') ] }`

Uniformity can already decrease complexity a lot.

**Comment by bebraw** [comment]
@andywer I agree that decreasing the amount of concepts to learn could be a worthwhile effort. Incidentally there's `webpack.SourceMapDevToolPlugin` already but not a lot of people know it exists. 😄

**Comment by andywer** [comment]
PS: Independent from that, the loader `test` regex really bugs me sometimes. It makes composing the webpack config from smaller config snippets nearly impossible.

Just an idea: Having a separate `file type regex => mime-type` mapping (given a set of sane default values most people will probably never have to touch it) and using the MIME type in the loader config might work well:

`loaders: [ { type: 'application/javascript', loaders: [ 'babel' ] } ]`

**Comment by bebraw** [comment]
@andywer Have you seen [concord](https://github.com/webpack/concord#package-configuration)? Mime based mapping might be in the cards. Not for webpack 2, though, but we'll see after that.

**Comment by andywer** [comment]
@bebraw No, but I'll check it out :)

**Comment by andywer** [comment]
@bebraw Yeah, [concord](https://github.com/webpack/concord#package-configuration) seems like a good concept to specify the MIME types. One could use this concord config if present or fall back to default types otherwise. Then use the MIME types in the module loader config to apply loaders based on file type, not regexs :)

**Comment by dmitriid** [comment]
It's awesome to see many people agree that strings as configs is a bad idea. However, it's not enough to just agree on the principle. Here's a piece of config extracted from `react-create-app`:

```
loaders:[{
   test: /(\.css)/,
   loader: ExtractTextPlugin.extract('style', 'css?-autoprefixer!postcss')
}]
```

This basically shows that every single other idea discussed in this topic is moot until the above goes away:

- you cannot split and merge configs
- you cannot validate configs
- you cannot get IDE autocompletion for such a config
- you cannot effectively build any tools on top of such a config

Basically this boils down to: unless these configs are simplified (apparently not before Webpack 3 at the earliest) no other proposal discussed here can be implemented.

Too bad it's probably too late to push for this for webpack 2.

**Comment by ericclemmons** [comment]
@dmitriid What's personally annoying about that example is that _the loader usage doesn't follow the other loader conventions_ (e.g. providing a `string` and a `query` object).

I don't know of a way to support it either:
> https://github.com/ericclemmons/terse-webpack/issues/27
(But I'm using `styled-componets` nowadays so :shrug: ?)

That's why my take is more along the lines of combing _what are you building_ with _here's the best way to build it_. (The user should only care about the former)

This is such a hard thing to do, do well, and it's a moonshot (IMO) to accomplish it, since removing complexity nearly always means removing functionality.

Still on the radar, but there's simply not a clear path/solution AFAIK... :(

**Comment by andywer** [comment]
> Still on the radar, but there's simply not a clear path/solution AFAIK... :(

Sure. But I think one can at least boil the previous comments down to two future improvements:
- Having fewer (and clear) concepts
- Do stuff consistently (I don't need a `loader` prop if there's a `loaders` prop which is a superset of the former, for instance)

I would hope that all the other issues will be easier to solve after improving these two basics.

**Comment by TheLarkInn** [comment]
Going to add some discussion I think is relevant to this discussion I came across on Twitter.
![image](https://cloud.githubusercontent.com/assets/3408176/20644039/a17da3c2-b3ed-11e6-97da-c32f535e7364.png)

Maybe these are some topics to challenge ourselves for designing this new interface for configuration

**Comment by TheLarkInn** [comment]
How do we feel about @andywer's webpack-blocks? This has gotten a lot of praise we should weigh up and downsides

**Comment by dmitriid** [comment]
Warning: 2 cents incoming :)

Looking at webpack-blocks, I can't really say it's a better solution:

- doesnt-do-anything-out-of-the-box-itis. Multiple configuration options spread over 10 different packages
- pulls everything into the upper level of configs: PostCSS configuration (which is a plugin config, unless I'm mistaken) is passed to the `createConfig` as a top-level configuration option on par with `entryPoint`
- easily leads to consistency abuse:
  - [extract-text](https://github.com/andywer/webpack-blocks/blob/master/packages/extract-text) is a function call with 0 or 2 arguments
  - [post-css](https://github.com/andywer/webpack-blocks/blob/master/packages/postcss) is a function that accepts an array of function calls (?)
  - [babel](https://github.com/andywer/webpack-blocks/blob/master/packages/babel6) is a function that accepts a single JavaScript object as options

I would opt for a config that in its base form is nothing but a declarative JavaScript object with consistent options everywhere:

So, for the above it should be something like:

```
{
  plugins: {   
    babel: {
        exclude : /node_modules/
    },
   postcss: {
     autoprefixer: {
       browsers: ['last 2 versions'] 
     }
   },
  extractText: {
    out: 'css/[name].css',
    mime: 'text/css'
  }
}
```

This way when a poorly written documentation for a plugin says "Pass the following options to plugin", the user will not have to try and guess how, where or when to pass the config options.

And yes, this will break a lot of plugins...

**Comment by dmitriid** [comment]
Also, unless I'm mistaken, gulp requires its plugins to be written in exactly one manner: that is 

```
import goodPlugin from 'goodPlugin'
import badPluginFactory from 'badPlugin'
const badPlugin = badPluginFactory(additional_config)

...
```

See https://github.com/gulpjs/gulp/blob/master/docs/writing-a-plugin/guidelines.md This is something webpack should strive for as well. At the very least it will make the plugin-related part of configuration simpler  and handling easier.

**Comment by andywer** [comment]
@dmitriid The arguments to the blocks are quite consistent, though, I think, *but* they might be more consistent nevertheless. But they got to reflect the different ways in which the loaders/plugins are configured.

But I don't even want to focus on webpack-blocks here. There is just one design decision that consistently gets me puzzled:

Why do I have to configure the Babel loader, for instance, as a string: `loader: "babel-loader?presets[]=es2015"`

As a webpack user I have got to serialize my parameters (could use the query object, but only if I use `loader`, not `loaders`), just so that in the end webpack got to require the `babel-loader` and deserialize the parameters again.

Why not do `loader: new BabelLoader({ presets: [ "es2015" ] })`? Otherwise it just feels like a detour through JSON land 😉

**Comment by joshhunt** [comment]
> Why do I have to configure the Babel loader, for instance, as a string: loader: "babel-loader?presets[]=es2015"

For what it's worth, this has been address in Webpack 2.x

```
{
  test: /\.(png|eot|svg|ttf|woff(2)?)/,
  loader: 'url-loader',
  query: {
    limit: 5 * 1024, // only inline if less than 5kb
    name: '[name]-[hash:6].[ext]'
  },
}
```

The changes to clean this up in Webpack 2 I think are much appreciated - the insane way of configuring loaders with config strings was the worst part of the config, and I'm glad to see that fixed.

Personally, I prefer declaring loaders and plugins as functions (like your example) rather than strings, but tbh I think that's more superficial and would have larger ecosystem impacts.

**Comment by dmitriid** [comment]
@joshhunt The `query` parameter was there in 1.x, too: http://webpack.github.io/docs/using-loaders.html#query-parameters

In webpack 2 it's renamed to [options](https://webpack.js.org/configuration/module/#useentry). Which is a goo thing :)

**Comment by dmitriid** [comment]
@andywer My reasoning for having everything in "JSON land" is this:

- all configuration is declarative
- all configuration is consistent: it's just the name of the thing we want to use followed by a key-value configuration

> loader: new BabelLoader({ presets: [ "es2015" ] })

BabelLoader might be a function that expects an obect. SomethingLoader might be a function that expects multiple arguments. SomethingElseLoader might be a function that expects an array of things. 

This results in an inconsistent mess (every loader's configuration looks and behaves differently) _and_ in an unmanageable config file (we can no longer merge, analyze, expand, serialize configs, see discussion above), as this basically comes back to

```
loader: ExtractTextPlugin.extract('style', 'css?-autoprefixer!postcss')
```

**Comment by baderbuddy** [comment]
What I would personally love to see is defaults in the loader's code. Instead of something like this.

`loaders: [
            { test: /\.css$/, loader: "style-loader!css-loader" }
        ]`

I could have something much more simple

`loaders: ['css']`

The css loader should be able to specify that the default test is `/\.css$/`, because 90% of the time that's what people are going to use it with, also it should be able to specify that the results of this loader by itself aren't generally useful, so a default prefix loader of style-loader is set.

Giving the loader creators that power to set these defaults could go a long way towards simplifying webpack configs.

**Comment by e-cloud** [comment]
@baderbuddy I think it would be better to create a wrapper for more general usage or create something like preset. Coz webpack is super feature-rich. It's hard to simplify the usage at the core level. Just like most linux command line tools.

**Comment by baderbuddy** [comment]
@e-cloud based on my usage I think if the loaders' maintainers set some default values and then webpack allowed the ability to overwrite them then that would probably make most people's usage easier. 

In my example the maintainer of the CSS loader can say that the default test is for the CSS extension and by default the values get fed into the style loader but someone using the loader could still specify all of those parameters to change the defaults to fit their needs.

I think webpack needs to be more opinionated in order to make it easier to jump from one webpack project to another. In my mind the best way to accomplish that is to let the loaders bring their own defaults for how to use them.

**Comment by TheLarkInn** [comment]
I know this is already a long standing thread but I think we should address discussions had here as part of our Usability overhaul. 

### require env as a parameter
https://twitter.com/dan_abramov/status/820045292032688133

I think there are pros and cons to this but it's valid to an extent.

**Comment by dmitriid** [comment]
Instead of opening an issue/feature request, I'll add to more things:

## 1. Loaders evaluation order 

Loaders should be evaluated left to right, top to bottom. The reason is simple: everything else in the config is evaluated this way.

Anywhere there's a list of things, they are evaluated FIFO-style. Only loaders are evaluated LIFO-style. This breaks principle of least surprise, requires context switching when reasoning about configuration etc. This is especially true now, when loaders are used within rules. The rules are evaluated left to right, top to bottom (FIFO). Once you arrive at a `Rule.loaders`, it's suddenly right to left, bottom to top (LIFO).

## 2. Plugins and loaders should have the same configuration style

Both plugins and loaders are some functions that are called with some options. However:

- loaders can be configured only declaratively by specifying the name of the loader and its options
- plugins can be configured only imperatively by importing/requiring a module, and building an object with required options

There's really no reason for plugins to be configured imperatively:

```
{
  plugins: [
    {
      plugin: 'uglifyjs-plugin',
      options: {
        mangle: true,
        sourceMap: false
      }
     },
     ...
  ]
}
```

This can be split/parsed/maintained/merged/analysed etc.

**Comment by TheLarkInn** [comment]
It's officially time that we start tackling this. With webpack 4 far undergoing its processes in webpack/webpack#next branch, its time that we at least have the planning and design specifications done for this milestone. 

I'm going to take this on as lead. I would also like 4-10 individuals who would love to be involved with the testing, feedback. @niieani @e-cloud @dmitriid you are all first ones that come to mind in terms of being extremely candid and pragmatic.

**Comment by dmitriid** [comment]
@TheLarkInn  I will say a preliminary yes. As always, life may intervene :(

**Comment by TheLarkInn** [comment]
No worries at all. After I discussed this a bit with @d3viant0ne, I realized there is one thing that blocks us for this:

1. Need to land our analytics (which will go live in webpack 4). 

However this shouldn't stop us from brainstorming ideas. 

On a shortlist: 

Progressive layering, zero config usability for multiple different assets, etc. but we prescribe the best practices. Then layering becomes, what kind of additions (maybe this is too much like babel presets?)

**Comment by d3viant0ne** [comment]
`Then layering becomes, what kind of additions (maybe this is too much like babel presets?)`

The babel presets were deprecated in favor of `babel-preset-env` where you set a target & it handles the rest.

In our case `webpack-preset-env` would be quite a bit more complex but the same approach would in theory be possible.

As far as entry level Webpack configs go, that could certain solve a lot of the usability gripes ( less specific config insistences mentioned by @dmitriid ).

You could then use a very familiar pattern to build on the default. i.e. 

```
const Configuration = {
  extends: ['@webpack/some-config', 'myPrivateConfig'], // the base webpack config has all the defaults

  rules: {
     ...Extend or override what you want here incrementally
  },
};

module.exports = Configuration;
```

Also allows people to build their own base configuration for those of us that want granular control of our entire configuration as well as the ability to publish a base configuration

**Comment by eliperelman** [comment]
Providing Neutrino as proof-of-concept of something that already has shown success: https://neutrino.js.org https://github.com/mozilla-neutrino/neutrino-dev

Neutrino provides a high-level API on top of webpack, in terms of making configuration extensible, predictable/deterministic, and easy to use through the use of presets and middleware.

A user can either use the CLI to use presets (e.g., `neutrino build --use @neutrinojs/react`), or drop a `.neutrinorc.js` file in their project (using the upcoming Neutrino v8 spec for reference):

```js
module.exports = {
  // Use a single piece of middleware
  use: [
    '@neutrinojs/react'
  ]
};
```

```js
module.exports = {
  // Use multiple pieces of middleware
  use: [
    '@neutrinojs/airbnb',
    '@neutrinojs/react',
    '@neutrinojs/jest'
  ]
};
```

It's all very Babel-like, so you can also set options for Neutrino in general, or specific options for presets/middleware:

```js
module.exports = {
  options: {
    output: 'dist' // instead of build/
  },
  use: [
    // override page title, which internally uses HtmlWebpack{Template/Plugin}
    ['@neutrinojs/react', { html: { title: 'My React App' } }]
  ]
};
```

Extensibility is exposed via the Neutrino configuration API, which uses [webpack-chain](https://github.com/mozilla-neutrino/webpack-chain), an alternate webpack configuration API we created so discreet configurations could still be modified in a predictable way, and completely informed by webpack's own configuration properties:

```js
module.exports = {
  use: [
    '@neutrinojs/react',
    // pass a function as middleware to use the Neutrino API
    (neutrino) => {
      // Change the babel configuration
      neutrino.config
        .rule('compile')
          .use('babel')
            .tap(options => { ...options, plugins: [...options.plugins, 'babel-plugin-whatever'] });
    }
  ]
};
```

This is something that is lacking with webpack configuration, because you cannot predict the nature of configuration when it is distributed across presets on npm. If you want to modify a Babel configuration, you must know the index of the rule within the config, you must know what format was used like rules or uses, and that this format isn't the same from one config to the next.

**Comment by ericclemmons** [comment]
Wow, that neutrino looks appealing & familiar (babel-esque).

I've done enough projects, though, that I'm ready to have my intuition proven wrong by the analytics coming in Webpack 4, and scoping from there.

**Comment by eliperelman** [comment]
@ericclemmons some of it should be *really* familiar https://github.com/mozilla-neutrino/neutrino-dev/tree/master/packages/start-server

**Comment by TheLarkInn** [comment]
Exactly. Once we land analytics, we can really carve out some hot paths and
make them relevant to our users. Nutrino is a great Nova Ars to look to.

**Comment by TheLarkInn** [comment]
Hey all, we've landed some changes into webpack 4  (IRT zero config, easier commons chunk defaults). I think these things are a great start but doesn't solve the whole story. I think presets could be the next best thing, but we'd like to land presets in either 4.x or 5. So I'm going to change the milestone and keep tracking as we are slowly trying to whiddle down requirements for 4.0 release! Thanks for keeping track of it also. And as always good discussion<3

**Comment by hackape** [comment]
Has anyone considered writing an IDE extension based on [./schemas/WebpackOptions.json](https://github.com/webpack/webpack/blob/71a12c8bc1fc06968c5028d03da261bf802f0c51/schemas/WebpackOptions.json)?

I really like the way VSCode handle its `setting.json`, as shown in screenshot:
![image](https://user-images.githubusercontent.com/8945315/39395112-38a8b036-4b0d-11e8-8d75-9d901d71a96b.png)

Basically it bakes the documentation into autocompletion and code hint, down to available values as enum and description for what each value does. As of IDE support, I think webpack should provide that too.  

Here's a [example](https://github.com/Microsoft/vscode/blob/29bead35f2a5339482158a7d48f7f0f3955629e2/src/vs/editor/common/config/commonEditorConfig.ts#L170) of how VSCode describe it's editor settings. As you can see, it's just JSON Schema, so this is viable.

**Comment by hackape** [comment]
@niieani mentioned utilizing typescript typings to help. I've tried the `@types/webpack`. I think it's nice to have TS linting and hinting on my webpack config, but it's not quite there. 

The missing block is, no description on value.
![image](https://user-images.githubusercontent.com/8945315/39395629-2117823e-4b13-11e8-9a80-723638940582.png)

TS type system cannot handle describing value, JSDoc style comment can only touch down to property level. Yet I think being able to provide hint a description on what each of the value does really makes the difference in webpack's case. JSON Schema can provide that level of granularity, and in fact we can generate TS typings basing on the JSON Schema. 

So as of IDE support, my thought is we should definitely write that IDE extension, and put some serious effort to maintain the JSON schema, make it a useful doc, maybe add link to descriptions, help people jump to the right place of the website.

**Comment by leodr** [comment]
IMO the rollup configuration file is a pretty great inspiration. It's so easy to add additional functionality with the plugins. The plugins can ship their own TypeScript types so their configuration is documented inside the editor, which is not possible with the current way of using strings for the webpack configuration like so:

```
{
    loader: 'postcss-loader',
    options: {
        plugins: () => [autoprefixer({ flexbox: 'no-2009' })]
    }
},
```
